pyro/infer/svgd.py

```diff
@@ -0,0 +1,215 @@
+import math
+from contextlib import ExitStack  # python 3
+
+import torch
+from torch.distributions import biject_to
+
+import pyro
+from pyro.distributions import Delta
+from pyro.infer.trace_elbo import Trace_ELBO
+from pyro.infer.autoguide.guides import AutoDelta
+from pyro.infer.autoguide.initialization import init_to_sample
+from pyro.distributions.util import sum_rightmost
+
+
+def vectorize(fn, num_particles, max_plate_nesting):
+    def _fn(*args, **kwargs):
+        with pyro.plate("num_particles_vectorized", num_particles, dim=-max_plate_nesting - 1):
+            return fn(*args, **kwargs)
+    return _fn
+
+
+class _SVGDGuide(AutoDelta):
+    """
+    This modification of :class:`AutoDelta` is used internally in the
+    :class:`SVGD` inference algorithm.
+    """
+    def __init__(self, model):
+        super(_SVGDGuide, self).__init__(model, prefix="svgd", init_loc_fn=init_to_sample)
+
+    def __call__(self, *args, **kwargs):
+        """
+        An automatic guide with the same ``*args, **kwargs`` as the base ``model``.
+
+        :return: A dict mapping sample site name to sampled value.
+        :rtype: dict
+        """
+        # if we've never run the model before, do so now so we can inspect the model structure
+        if self.prototype_trace is None:
+            self._setup_prototype(*args, **kwargs)
+
+        plates = self._create_plates()
+        result = {}
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            with ExitStack() as stack:
+                for frame in site["cond_indep_stack"]:
+                    if frame.vectorized:
+                        stack.enter_context(plates[frame.name])
+                value = pyro.param("{}_{}".format(self.prefix, name),
+                                   site["value"].detach(),
+                                   constraint=site["fn"].support)
+                # The following lines are where this method differs from the AutoDelta version.
+                # For SVGD (but not MAP) we need the log det jacobian terms for correctness.
+                transform = biject_to(site["fn"].support)
+                unconstrained_value = pyro.param("{}_{}".format(self.prefix, name)).unconstrained()
+                log_density = transform.inv.log_abs_det_jacobian(value, unconstrained_value)
+                log_density = sum_rightmost(log_density, log_density.dim() - value.dim() + site["fn"].event_dim)
+                result[name] = pyro.sample(name, Delta(value, log_density=log_density,
+                                           event_dim=site["fn"].event_dim))
+        return result
+
+
+class SteinKernel(object):
```

Comment:
TODO: make this into an abstract base class with an `@abstractmethod` `.kernel_and_grads()`.